mlir/lib/Dialect/Tensor/IR/TensorOps.cpp

```diff
@@ -3791,13 +3792,78 @@ struct FoldConsecutiveConstantPadding : public OpRewritePattern<tensor::PadOp> {
   }
 };
 
+struct FoldReifiedShape : public OpRewritePattern<tensor::PadOp> {
+  using OpRewritePattern<tensor::PadOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(tensor::PadOp padOp,
+                                PatternRewriter &rewriter) const override {
+    if (padOp.getNofold()) {
+      return rewriter.notifyMatchFailure(padOp, "skipping unfoldable pad");
+    }
+
+    ReifiedRankedShapedTypeDims reifiedResultShapes;
+    if (failed(reifyResultShapes(rewriter, padOp, reifiedResultShapes)))
+      return failure();
```

Comment:
This is potentially modifying the IR before we know we can apply.

Reply:
Yes, unfortunately `reifyResultShapes` cannot be used like that in a canonicalization pattern. You don't want the pattern to apply when the reified shape is the same. But in order to see the reified shape, you have generate IR first, at which point you can no longer abort.